### PR TITLE
SOLR-15616: Allow thread metrics to be cached

### PR DESCRIPTION
**Pull request by chatman**
Details are here: https://issues.apache.org/jira/browse/SOLR-15616
Tested manually by setting the sysprop in bin/solr.in.sh and hitting the metrics API periodically.

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java [hunk 1]
```diff
@@ -213,7 +216,14 @@ private void setupJvmMetrics(CoreContainer coresInit)  {
       metricManager.registerAll(registryName, new OperatingSystemMetricSet(), SolrMetricManager.ResolutionStrategy.IGNORE, "os");
       metricManager.registerAll(registryName, new GarbageCollectorMetricSet(), SolrMetricManager.ResolutionStrategy.IGNORE, "gc");
       metricManager.registerAll(registryName, new MemoryUsageGaugeSet(), SolrMetricManager.ResolutionStrategy.IGNORE, "memory");
-      metricManager.registerAll(registryName, new ThreadStatesGaugeSet(), SolrMetricManager.ResolutionStrategy.IGNORE, "threads"); // todo should we use CachedThreadStatesGaugeSet instead?
+      if (System.getProperties().contains("solr.metrics.threads.caching.interval")) {
+        int threadsCachingInterval = Integer.valueOf(System.getProperty("solr.metrics.threads.caching.interval"));
+        log.info("Threads metrics will be cached for " + threadsCachingInterval + " seconds");
+        metricManager.registerAll(registryName, new CachedThreadStatesGaugeSet(threadsCachingInterval, TimeUnit.SECONDS), SolrMetricManager.ResolutionStrategy.IGNORE, "threads");
```
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java]: is it possible to get only thread_count and runnable_thread?

File: solr/solr-ref-guide/src/metrics-reporting.adoc [hunk 2]
```diff
@@ -211,6 +211,10 @@ complex objects:
 </metrics>
 ----
 
+=== Caching Threads Metrics ===
+The threads metrics in the JVM group can be expensive to compute, as it requires traversing all threads. 
+This can be avoided for every call to the metrics API (group=jvm) by setting a higher caching expiration interval (in seconds)
+by passing a system property, e.g. `solr.metrics.threads.caching.interval=5` (to cache the thread metrics for 5 seconds).
```
Comment (hiteshk25) [inline solr/solr-ref-guide/src/metrics-reporting.adoc:217]: we make call every 15 seconds.

File: solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java (hunk 1 above)
Comment (chatman) [reply solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java]: > is it possible to get only thread_count and runnable_thread?

I've updated the PR to disable the deadlock detection with a no-op impl, that should reduce the computation now. When this is disabled, threads.deadlocks.* metrics will be zero/empty.

**Comment by chatman** [comment]
I think we should just disable the deadlock metrics with this sysprop:
`-Dsolr.metrics.threads.deadlockdetection.enabled=false`
and not bother with caching etc.

File: solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java [hunk 3]
```diff
@@ -227,6 +234,44 @@ private void setupJvmMetrics(CoreContainer coresInit)  {
     }
   }
 
+  class NoOpThreadsDeadlockDetector extends ThreadDeadlockDetector {
```
Comment (noblepaul) [inline solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java:237]: can this be put into a separate class. `SolrDispatchFilter` is already unnecessarily complex

File: solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java [hunk 4]
```diff
@@ -227,6 +234,44 @@ private void setupJvmMetrics(CoreContainer coresInit)  {
...
+    public Set<String> getDeadlockedThreads() {
+      return Collections.emptySet();
+    }
+  }
+
+  private void registerThreadsMetrics() throws Exception {
+    boolean threadsDeadlockDetectionEnabled = 
+        System.getProperties().contains("solr.metrics.threads.deadlockdetection.enabled")? 
+            Boolean.valueOf(System.getProperty("solr.metrics.threads.deadlockdetection.enabled"))
+            : false; // enabled by default, unless this sysprop is set to false
```
Comment (hiteshk25) [inline solr/core/src/java/org/apache/solr/servlet/SolrDispatchFilter.java:248]: so we want to set default as true?

**Comment by hiteshk25** [comment]
@chatman @noblepaul we will just disable deadlock detection for now. is that fine?
```
-Dsolr.metrics.threads.deadlockdetection.enabled=false
```

**Review by hiteshk25 — APPROVED** [review]
LGTM

**Pull request merged**